### DISCUSSION TITLE
how to add mcu to cmsis_mcu_descriptions.json5

### DISCUSSION THREAD

**Discussion by jancotipsarevic** [mbed-ce/mbed-os#389] [General]
Hello, I am planing to use a board with STM32F103VGT6 but it is not listed in the cmsis_mcu_descriptions.json5 file and complains running cmake ..
"ERROR: Target specifies device_name STM32F103xG but this device is not
listed in C:\Mbed-CE\mbed-ce-hello-world\mbed-os\tools\cmake\..\..\targets\cmsis_mcu_descriptions.json5.  Perhaps you need to use
the 'python -m mbed_tools.cli.main cmsis-mcu-descr fetch-missing' command to download
the missing MCU description?"

How is the proper way to create this entry in the json5 file?

Thank you very much

**Comment by timIdeaTech** [comment]
As it happens, I ran into the the same issue yesterday too. As it turns out STM32L433CCTx doesn't exist in cmsis_mcu_descriptions.json5 either. I'm porting an old project from mbed-os to mbed-ce.

If I'm correct: the mbed-ce fork of mbed-os, mbed-os/targets/cmsis_mcu_descriptions.json5 is the same as the mbed-ce forked version of the legacy mbed-os/tools/arm_pack_manager/index.json. It would be easy to say we should just update cmis_mcu_descriptions.json5 in mbed-ce but it may not be that easy.

So I add to the question: can the above update be accomplished with out breaking something or is there a way to add to the mcu descriptions, at least temporarily, using let's say the custom_target or some other method similar to the mbed_disable_mcu_target_file() cmake call for source files?

**Reply by jancotipsarevic** [reply]
It is a good point, I will give it a try just to adding the missing entry for the processor I plan to use (harcoded in the cmsis_mcu_descriptions.json5) to see what happens. Thank you very much for your tip.

**Reply by JohnK1987** [reply]
Hello guys,
Sorry I am late. I did not try it by my self because I do not use any custom target right now but I will try to point somewhere.

First  the `python -m mbed_tools.cli.main cmsis-mcu-descr fetch-missing` scanning just targets.json5 file and not custom_targets.json5 file.
But when we look [here](https://github.com/mbed-ce/mbed-os/wiki/Mbed-Memory-Bank-Information#33-interaction-with-custom-targets) and then [here](https://github.com/mbed-ce/mbed-ce-custom-targets/blob/ac67ef1fb69f0ebea482b862b0e7cbb5cbc5ac37/custom_targets/custom_targets.json5#L24-L86), then it looks like you have to add `memory_banks` sections to your `custom_targets.json5` file. This will probably skip the `cmsis-mcu-descr` requirements.

**Reply by timIdeaTech** [reply]
Adding the `memory_banks` section to `custom_targets.json5` is the quick and/or temporary way to get what we want here. I should have looked deeper into the latest `mbed-ce-custom-targets` example, it's laid out there. I missed the comment about "should NOT specify a device_name", which now makes sense when adding this section.

Regarding updating the `cmsis-mcu-descr` file in mbed-ce, do you see that being useful or being a possibility?

While we are on the subject of device memory maps, is it possible to alter individual keys in the `memory_banks` section or do you need to replicate the entire record in the `custom_targets.json5` file? The key values eventually become defines so maybe that is another method to alter individual keys, just as long as the macro is set in the right sequence that cmake puts it all together.

**Reply by multiplemonomials** [reply]
I would accept a PR that would allow passing a `custom_mcu_descriptions.json5` file into the mbed-tools python scripts.  Not sure if I have time right now to do that though, and there is no difference in functionality vs using the memory_banks attribute as far as I'm aware.

How it works is, any memory banks defined in a target in custom_targets.json will replace memory banks of the same name defined from the parent target, whether those come from device_name/cmsis_mcu_descriptions or they are declared inline.  So, for example, if I extend a target from Mbed that already has a device_name which pulls in memory banks information, I can replace the definition for one specific memory bank by declaring it in `memory_banks` for my custom target.  The same would be true if the parent target just had a `memory_banks` attribute that declared the memory banks.

Hope that makes sense!

**Reply by timIdeaTech** [reply]
I'm not sure I follow your PR comment. Are you saying there is an mbed-ce script to generate, update or otherwise get custom_mcu_descriptions.json5 up to date with the latest legacy `mbed-os/tools/arm-pack-manager/*` files?

I don't see a downside to updating `custom_mcu_descriptions.json5` from the lastest legacy mbed-os `index.json` file to capture the legacy updates since the mbed-ce fork. This assumes of course that they are essentially the same files as I surmised and no changes have been done to `custom_mcu_descriptions.json5` since the fork (I checked the repo and I'm pretty sure there has been but some/all could be already updated in legacy).

I did a compare (WinMerge) of the two files and there appears to be a large number of additions to `index.json`, to the point the tool has a problem sync'ing up. I also don't know how the legacy `aliases.json` and `legacy-names.json` files fit/existed/taken care of when the mbed-ce fork was created.

Yes, adding memory_banks entries into the `custom_targets.json5` solves the problem per new project. Is there a drive to capture as much of the legacy 6.17 changes to mbed-ce before it ARM kills the site?

**Reply by timIdeaTech** [reply]
Sorry Jamie, brain fart. I get your PR comment now as you'll notice I misread the `custom_mcu...` for `cmsis_mcu...` and even included the bf in my reply :^/

My vote in this case would to be keep it simple and use the memory_banks in custom_targets.json5 method. Again, that method is already been laid out in the `mbed-ce-custom-targets` example.

fwiw, about the change of `memory_banks` key value, `IROM1` in this case, it looks like you can change them in a `memory_banks` entry in `custom_targets.json5` but you have to enter the whole `IROM1` branch rather than just one end key:value pair. You might be wondering why I'm looking at this. It has to do with porting a the legacy mbed-os bootloader project, which as far as I can determine isn't supported by by mbed-ce. Feel free to correct me if I'm wrong.

Cheers!

**Reply by jancotipsarevic** [reply]
Hello all , in my case it looks it is not just a matter of memories definition, the part STM32F103VG is not listed in cmsis_mcu_descriptions.json.
When I copy the STM32F103VG entry from "tools/arm_pack_manager/index.json" to cmsis_mcu_descriptions.json everything works fine. 

This is the error that appears even if I put the memory banks in my custom_tagets.json
ERROR: Target specifies device_name STM32F103VG but this device is not

My custom_targets.json (working, with no memory_banks as it looks it is required anyway in cmsis_mcu_descriptions.json) 
{
    "MYBOARD": {
        "inherits": [
            "MCU_STM32F103xG"
        ],
        "device_name": "STM32F103VG",
        "macros_add": [
            "HSE_VALUE=16000000"
        ],
        "overrides": {
            "clock_source": "USE_PLL_HSE_XTAL",
            "lse_available": 1
        }
    }
}

**Reply by multiplemonomials** [reply]
> I did a compare (WinMerge) of the two files and there appears to be a large number of additions to index.json, to the point the tool has a problem sync'ing up. I also don't know how the legacy aliases.json and legacy-names.json files fit/existed/taken care of when the mbed-ce fork was created.

Yeah, I was unhappy about how big that JSON file was in the Mbed OS source tree.  It contained definitions for basically every ARM CPU in the CMSIS index, instead of just the ones used by Mbed.  I wrote a script to pare it down to just the ones that were actually used in the current version of Mbed.  That way it's smaller and only has the MCUs we actually use and can maintain.

Note that if we could just use the CMSIS descriptions unmodified, we wouldn't have to have them stored in a JSON file in the source tree, we could just download them when we need them.  Unfortunately, a number of the CMSIS descriptions have significant issues, and have to be hand edited before they are usable in Mbed (for instance, the STM32H743 MCU description is missing DTCM and backup SRAM, and the MIMXRT1052 description has the factory-burned, unprogrammable ROM listed as flash!).

**Reply by multiplemonomials** [reply]
> You might be wondering why I'm looking at this. It has to do with porting a the legacy mbed-os bootloader project, which as far as I can determine isn't supported by by mbed-ce. Feel free to correct me if I'm wrong.

That's really awesome to hear!  I would love to see a return of the bootloader project to Mbed.  This is actually a significant reason why I added the memory banks feature in the first place: I wanted to restore the ability to configure the flash area so that Mbed CE could properly coexist with a bootloader.  Unfortunately there are still a lot more linker script updates to do before that can happen though.

**Reply by JohnK1987** [reply]
@jancotipsarevic 
> Hello all , in my case it looks it is not just a matter of memories definition, the part STM32F103VG is not listed in cmsis_mcu_descriptions.json. When I copy the STM32F103VG entry from "tools/arm_pack_manager/index.json" to cmsis_mcu_descriptions.json everything works fine.
> 
> This is the error that appears even if I put the memory banks in my custom_tagets.json ERROR: Target specifies device_name STM32F103VG but this device is not
> 
> My custom_targets.json (working, with no memory_banks as it looks it is required anyway in cmsis_mcu_descriptions.json) { "MYBOARD": { "inherits": [ "MCU_STM32F103xG" ], "device_name": "STM32F103VG", "macros_add": [ "HSE_VALUE=16000000" ], "overrides": { "clock_source": "USE_PLL_HSE_XTAL", "lse_available": 1 } } }

You do it almost right but how was mentioned above
>I missed the comment about "should NOT specify a device_name", which now makes sense when adding this section.

So try to remove `device_name` from your custom_targets.json5`.

**Reply by jancotipsarevic** [reply]
Sorry, now works, thank you very much.

**Reply by timIdeaTech** [reply]
@multiplemonomials 
> > You might be wondering why I'm looking at this. It has to do with porting a the legacy mbed-os bootloader project, which as far as I can determine isn't supported by by mbed-ce. Feel free to correct me if I'm wrong.
> 
> That's really awesome to hear! I would love to see a return of the bootloader project to Mbed. This is actually a significant reason why I added the memory banks feature in the first place: I wanted to restore the ability to configure the flash area so that Mbed CE could properly coexist with a bootloader. Unfortunately there are still a lot more linker script updates to do before that can happen though.

I actually ran into #305 while looking at this and realized there is more to this than I expected. My questions stemmed from an attempt to provide a temporary but quick solution to the port of our legacy project to mbed-ce and see if and where we were going to run into issues any other issues.

I also ran into various comments about MCUboot maybe being a better option than the mbed-os bootloader. Is there a consensus on which method the mbed-ce will gravitate too?

**Reply by JohnK1987** [reply]
@jancotipsarevic 
> Sorry, now works, thank you very much.

Do not worry, that is why we have discussions here.

@timIdeaTech I appreciate your activity here and do not understand me wrong but maybe a new topic about the bootloader will be better place for discussion.

**Reply by JohnK1987** [reply]
I made small update of `fetch_missing` in `cmsis_mcu_descr.py`. It is able help also with custom targets now - https://github.com/mbed-ce/mbed-os/pull/394

**Reply by timIdeaTech** [reply]
@JohnK1987 
No worries. Related but yes a new, #305 or other discussion thread. I wandered because of the `memory_banks` link to what I was trying to quickly accomplish.
